cmd/maya-exporter/app/collector/collector.go

```diff
@@ -122,9 +137,21 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		stats       stats
 	)
 
+	c.Lock()
+	if c.isRequestInProgress() {
+		c.targetRejectRequestCounter.Inc()
+		c.targetRejectRequestCounter.Collect(ch)
+		c.Unlock()
+		return
+	}
+
+	c.request = true
+	c.Unlock()
+
 	klog.V(2).Info("Get metrics")
 	metrics := &c.metrics
 	if volumeStats, err = c.get(); err != nil {
+		c.setRequestToFalse()
```

Comment:
Do we need to unset request here? If one call errored and second one got succeeded then final metrics may not have expected value in both call. WDYT?

Reply:
@mynktl other request will be succeeded only after resetting it, since lock is only protecting the c.request flag. In case of error we want that other blocked requests continue as early as possible.

Reply:
in other words, until we reset the flags, other requests will be rejected here in this piece of code
```
	c.Lock()
	if c.isRequestInProgress() {
		c.targetRejectRequestCounter.Inc()
		c.targetRejectRequestCounter.Collect(ch)
		c.Unlock()
		return
	}
```

Reply:
these function will get executed in go routine so we can't be sure of the sequence. Consider following:
routine 1:
c.get() returned error and set request false
routine 2:
executed c.get() since routine 1 set the request to false.

now in both routine are trying to assign stats to c.metrics, one has empty stats and other one has proper. We can't be sure if second routine assigned first or after routine 1.